R/exec_profile.R

```diff
@@ -0,0 +1,57 @@
+exec_profile <- function(.fn, indicator, indicator_after) {
+  tilt_indicator_output <- exec(get(.fn), !!!indicator)
+
+  product <- unnest_product(tilt_indicator_output) |>
+    extend_with_columns_from_arguments_of_tilt_indicator(indicator, .fn)
+  company <- unnest_company(tilt_indicator_output)
+
+  out_product <- exec(polish_product(.fn), product, !!!indicator_after)
+  out_company <- exec(polish_company(.fn), company, product, !!!indicator_after)
+
+  nest_levels(out_product, out_company)
+}
+
+extend_with_columns_from_arguments_of_tilt_indicator <- function(product,
+                                                                 indicator,
+                                                                 .fn) {
+  if (grepl("sector_profile$", .fn)) {
+    companies <- indicator[[1]]
+    out <- product |> extend_with(companies)
+  }
+  if (grepl("emissions", .fn)) {
+    co2 <- indicator[[2]]
+    out <- product |> extend_with(co2)
+  }
+  if (grepl("sector_profile_upstream$", .fn)) {
+    inputs <- indicator[[3]]
+    out <- product |> extend_with(inputs)
+  }
+
+  out
+}
+
+extend_with <- function(data, with, cols_pattern = extra_cols_pattern()) {
+  .with <- with |> select(matches(cols_pattern))
+  .data <- data |> select(-any_of(names(.with)), matches("rowid"))
+  left_join(.data, .with, by = extra_rowid())
+}
```

Comment:
I'm now sure what your implementation would look like but if it's clear to you then go ahead and push the refactoring. 

It was a pretty tricky part and this implementation seems to do what we need but I may be wrong. Note the columns are excluded from the `data1` to avoid conflicts when joining with `data2`, since we want the columns from `data2` anywyay.

``` r
library(dplyr, warn.conflicts = FALSE)

data1 <- tibble(extra_rowid = 1, tilt_sector = 1, other1 = 1, tilt_subsector = 1)
data2 <- tibble(extra_rowid = 1, tilt_sector = 2, other2 = 2)

# Bad we get the suffix .x and .y
left_join(data1, data2, by = "extra_rowid")
#> # A tibble: 1 × 6
#>   extra_rowid tilt_sector.x other1 tilt_subsector tilt_sector.y other2
#>         <dbl>         <dbl>  <dbl>          <dbl>         <dbl>  <dbl>
#> 1           1             1      1              1             2      2

extra_pattern <- c("rowid", "sector")
.extension <- data2 |> select(matches(extra_pattern))
.extension
#> # A tibble: 1 × 2
#>   extra_rowid tilt_sector
#>         <dbl>       <dbl>
#> 1           1           2

.data <- data1 |> select(-any_of(names(.extension)), "extra_rowid")
.data
#> # A tibble: 1 × 3
#>   other1 tilt_subsector extra_rowid
#>    <dbl>          <dbl>       <dbl>
#> 1      1              1           1

left_join(.data, .extension, by = "extra_rowid")
#> # A tibble: 1 × 4
#>   other1 tilt_subsector extra_rowid tilt_sector
#>    <dbl>          <dbl>       <dbl>       <dbl>
#> 1      1              1           1           2
```

<sup>Created on 2023-12-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

-- 

I need to go for a few hours now. Feel free to take it over.

Reply:
@maurolepore Understood! Now, I don't think any change would be required. Thanks for the reprex!